src/include/txn_inline.h

```diff
@@ -1163,11 +1163,7 @@ __wt_txn_read_upd_list_internal(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt,
          */
         if (upd->type == WT_UPDATE_TOMBSTONE && F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE) &&
           !WT_TIME_WINDOW_HAS_STOP(&cbt->upd_value->tw)) {
-            cbt->upd_value->tw.durable_stop_ts = upd->durable_ts;
-            cbt->upd_value->tw.stop_ts = upd->start_ts;
-            cbt->upd_value->tw.stop_txn = upd->txnid;
-            cbt->upd_value->tw.prepare =
-              prepare_state == WT_PREPARE_INPROGRESS || prepare_state == WT_PREPARE_LOCKED;
+            __wt_upd_value_assign(cbt->upd_value, upd);
```

Comment:
I dislike adding a second `WT_ACQUIRE_READ_WITH_BARRIER` for `prepare_state`. It's unavoidable since we need `__wt_upd_value_assign` to use it for correctness, but we also need to check `prepare_state` against `WT_PREPARE_RESOLVED` above. Besides not being free (performance-wise), this second read also raises the possibility of the value changing between the two reads.

It may not be a problem that the value changes between the reads, or it may not be possible. But either way it's harder to reason about than just leaving the original three lines intact.

So for that reason, I'm inclined to NAK this part of the change - I think the cleanup inside `__wt_upd_value_assign` is useful though. What do you think, @luke-pearson?

Reply:
We do call `__wt_upd_value_assign` later in this function though. The real fix here would be to mace prepare state an argument to that function making it explicit that the caller needs to handle the ordering semantics.